### DISCUSSION TITLE
Colouring groups of `$status_format` can only sometimes overlap

### DISCUSSION THREAD

**Discussion by whitney-cumber** [neomutt/neomutt#4521] [Development]
## Expected Behaviour

If not allow arbitrary overlaps, at least I expect consistent overlapping rules.

## Actual Behaviour

Colouring groups of `$status_format` can only sometimes overlap.

## Steps to Reproduce

Consider the following `color_neomuttrc`

```
set status_format = "aaabbbccc"

# Case 1a: Cannot colour overlaps
#color status red white "aaabbbccc"
#color status blue white "bbb"

# Case 1b: Cannot colour overlaps (order does not matter)
color status blue white "bbb"
color status red white "aaabbbccc"

# Case 2: Cannot match overlaps
#color status red white "aaabbb"
#color status blue white "bbb(ccc)" 1

# Case 3: Cannot overlap matches
#color status red white "aaabbb(ccc)" 1
#color status blue white "bbb"

# Case 4: Can match non-coloured overlaps
#color status red white "(aaa)bbb" 1
#color status blue white "bbb(ccc)" 1

# Case 5a: Can match overlaps "to the right"
#color status red white "(aaa)bbb" 1
#color status blue white "bbbccc"

# Case 5b: Cannot match overlaps "to the left"
#color status red white "aaabbb"
#color status blue white "bbb(ccc)" 1

```

(Obviously, uncomment the case you want to see.)
Then run `neomutt -F color_neomuttrc` and look at the status bar.

Case 1: Overlaps of matches are not permitted. Case 1b demonstrates that the order of the `color` commands seems to be insignificant.
*Result:* Everything is red.

Case 2: Some overlaps do not work.
*Result:* `aaabbb` is red, `ccc` is uncoloured.

Case 3: Even though "aaabbbccc" matches first, "bbb" wins, in contrast to case 1.
*Result:* `bbb` is blue, `aaa`, `ccc` is uncoloured.

Case 4: Sometimes overlaps work!? (Compare to Case 2)
*Result:* `aaa` is red, `bbb` uncoloured, `ccc` blue.

Case 5: Case 5a works even though the symmetric case 5b does not.
*Result 5a*: `aaa` red, `bbbccc` blue.
*Result 5b*: `aaabbb` red, `ccc` uncoloured.

## Wish/Idea

Best would be to allow overlaps with one of the rules "last match wins" or "first match wins" (similar to how hook patterns work).

## How often does this happen?

- Always

## NeoMutt Version

```
20220429
```

**Comment by flatcap** [comment]
I'll test your examples, when I have a second, later.

IIRC, NeoMutt works through the status string one char at a time...

If there are multiple matches for that position, the longest one wins.
If there are multiple same-length matches, the last configured one wins.
Then (I think) NeoMutt skips the length of the match and continues.

**Comment by flatcap** [comment]
Yep, your examples and conclusions are spot on.
(and the feature is behaving as I expected :-)

I recently overhauled the colour management so that colours were applied more consistently.
A colour of `default` now means 'transparent', so that 'red default' + 'default blue' == 'red blue'.
Except that I didn't alter the status colour feature.  Not sure why not.

The design of the feature meant that we didn't have to worry about overlaps or merging.

> Case 1a: Cannot colour overlaps
> color status red white "aaabbbccc"
> color status blue white "bbb"

Yes

Internally, we'd create the 'red' colour for the entire span,
then have to split it in three and merge the 'blue' into middle object.

> Case 2: Cannot match overlaps
> color status red white "aaabbb"
> color status blue white "bbb(ccc)" 1

So rather than skipping the matched part, we just move to the next char.

---

Yes, we should do this, but the work'll be quite involved.
Most of the code we'd need is already there, in the index, etc.

We just need a volunteer...
